<a href="https://colab.research.google.com/github/Sagarjain93/car_prices/blob/main/car_prices.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Project Title : Exploratory Data Analysis on Sales of Car Dataset**

##**1.Introduction**

This Exploratory Data Analysis (EDA) provides a clear overview of the car sales dataset, which includes important details such as car brand, model, manufacturing year, price, fuel type, transmission, and mileage. The primary goal is to uncover actionable insights that help us understand customer preferences, market trends, and factors driving sales performance.

Through careful data cleaning and visualization, we highlight key patterns and relationships that influence buying decisions and pricing. These insights will support strategic business decisions, improve sales forecasting, and identify opportunities for growth in a competitive automotive market. Ultimately, this analysis forms a data-driven foundation for optimizing marketing strategies and enhancing overall business performance.

we hypothesize that

  1.Higher car mileage (odometer) leads to lower selling prices.
  Rationale: Cars with more wear and tear are typically valued less.

  2.Cars in better condition sell for higher prices.
  Rationale: Buyers are willing to pay more for well-maintained vehicles.

  3.Certain brands (e.g., Toyota, Honda) retain value better than others.
  Rationale: Some brands are known for reliability and resale value.

  4.Automatic transmission cars sell for higher prices than manual ones.
  Rationale: Automatic vehicles are more popular and convenient for most buyers.

  5.Newer model years are associated with higher selling prices.
  Rationale: Newer vehicles have updated features and less wear.

  6.MMR (market value estimate) is positively correlated with actual selling price.
  Rationale: MMR is a pricing benchmark and should closely match real transactions.

  7.SUVs and trucks have higher average selling prices than sedans and hatchbacks.
  Rationale: Larger vehicles are generally more expensive.

  8.Sales volume or price varies by state due to demand or regional preferences.
  Rationale: Certain states may have more active used car markets or different price trends.

  9.Cars sold in winter months have lower prices compared to summer months.
  Rationale: Consumer demand may fluctuate seasonally.

  10.Dealer (seller) has an impact on selling price even for similar models.
  Rationale: Some dealers may offer better deals or add premiums.

##**2.Dataset Description**

The car sales dataset contains detailed information on vehicle listings and sales transactions. The key attributes included are:

* **Year:** The model year of the vehicle.

* **Make:** The manufacturer or brand of the car (e.g., Toyota, Ford).

* **Model:** The specific model name of the vehicle.

* **Trim:** The variant or version of the model, indicating different features or packages.

* **Body:** The body style of the vehicle (e.g., sedan, SUV, coupe).

* **Transmission:** The type of transmission system (e.g., automatic, manual).

* **VIN:** The unique Vehicle Identification Number, used to track individual cars.

* **State**: The U.S. state where the vehicle was sold or listed.

* **Condition**: The overall condition rating of the vehicle (e.g., excellent, good, fair).

* **Odometer:** The mileage on the vehicle at the time of sale, measured in miles.

* **Color**: The exterior color of the vehicle.

* **Interior**: The interior color or material of the vehicle.

* Seller: The entity or dealership selling the car.

* **MMR (Manheim Market Report)**: The wholesale market value estimate for the vehicle.

* **Selling Price:** The final sale price of the vehicle.

* **Sale Date:** The date when the sale transaction occurred.

This dataset provides a comprehensive view of car sales transactions, enabling analysis of price trends, vehicle popularity, market conditions, and sales performance across different regions and time periods.



#**3. Import Required Libraries**
We import Python libraries necessary for data manipulation and visualization

In [1]:
# Data Manipulation Libraries
import numpy as np
import pandas as pd

# Data Visualization Libraries
import matplotlib.pyplot as plt
import seaborn as sns

# Set a consistent theme for all plots
sns.set(style = 'whitegrid')

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#**4. Load The Dataset**
To begin the analysis, the car sales dataset is loaded directly from Google Drive. This approach ensures convenient access to the data stored in the cloud, especially when working in collaborative or cloud-based environments like Google Colab. By mounting Google Drive or using a shareable link, we can seamlessly import the dataset into our workspace for further processing and analysis.



In [3]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/6_Sales/car/car_prices.csv')


#**5. Initial Data Inspection**
In this step, we perform an initial examination of the dataset to understand its structure and quality. This includes previewing sample records, checking the dataset’s size, identifying data types, and detecting missing or duplicate values. These insights help inform subsequent data cleaning and analysis steps.

#**5.1. Preview First Few Records**
Display the first 5 rows to get an initial idea of the dataset.

In [4]:
df.head()

year   make                model        trim   body transmission  \
0  2015    Kia              Sorento          LX    SUV    automatic   
1  2015    Kia              Sorento          LX    SUV    automatic   
2  2014    BMW             3 Series  328i SULEV  Sedan    automatic   
3  2015  Volvo                  S60          T5  Sedan    automatic   
4  2014    BMW  6 Series Gran Coupe        650i  Sedan    automatic   

                 vin state  condition  odometer  color interior  \
0  5xyktca69fg566472    ca        5.0   16639.0  white    black   
1  5xyktca69fg561319    ca        5.0    9393.0  white    beige   
2  wba3c1c51ek116351    ca       45.0    1331.0   gray    black   
3  yv1612tb4f1310987    ca       41.0   14282.0  white    black   
4  wba6b2c57ed129731    ca       43.0    2641.0   gray    black   

                                   seller      mmr  sellingprice  \
0                 kia motors america  inc  20500.0       21500.0   
1                 kia motors america  inc  20800.0       21500.0   
2  financial services remarketing (lease)  31900.0       30000.0   
3                 volvo na rep/world omni  27500.0       27750.0   
4  financial services remarketing (lease)  66000.0       67000.0   

                                  saledate  
0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)

#**5.2. Check The Dataset Shape**
Check the number of rows and columns in the datase

In [5]:
df.shape

(558837, 16)

#**5.3 Dataset Summary Overview**
Check for missing values and data types of each column.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558837 entries, 0 to 558836
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558837 non-null  int64  
 1   make          548536 non-null  object 
 2   model         548438 non-null  object 
 3   trim          548186 non-null  object 
 4   body          545642 non-null  object 
 5   transmission  493485 non-null  object 
 6   vin           558833 non-null  object 
 7   state         558837 non-null  object 
 8   condition     547017 non-null  float64
 9   odometer      558743 non-null  float64
 10  color         558088 non-null  object 
 11  interior      558088 non-null  object 
 12  seller        558837 non-null  object 
 13  mmr           558799 non-null  float64
 14  sellingprice  558825 non-null  float64
 15  saledate      558825 non-null  object 
dtypes: float64(4), int64(1), object(11)
memory usage: 68.2+ MB


**Interpretation** The car sales dataset includes 558,837 records with 16 columns detailing vehicle specs like year, make, model, body type, transmission, mileage, condition, and selling price. Most key fields are complete, though transmission, body, and trim have moderate missing values. The saledate column needs conversion to datetime for temporal analysis. Overall, the dataset is well-structured for exploring pricing trends, brand performance, and sales patterns across U.S. states.

#**5.4. Statistical Summary of Numeric Columns**
Generate descriptive statistics such as mean, standard deviation, min, max, and quartiles for numerical columns.

In [7]:
df.select_dtypes(include='number').describe()

year      condition       odometer            mmr  \
count  558837.000000  547017.000000  558743.000000  558799.000000   
mean     2010.038927      30.672365   68320.017767   13769.377495   
std         3.966864      13.402832   53398.542821    9679.967174   
min      1982.000000       1.000000       1.000000      25.000000   
25%      2007.000000      23.000000   28371.000000    7100.000000   
50%      2012.000000      35.000000   52254.000000   12250.000000   
75%      2013.000000      42.000000   99109.000000   18300.000000   
max      2015.000000      49.000000  999999.000000  182000.000000   

        sellingprice  
count  558825.000000  
mean    13611.358810  
std      9749.501628  
min         1.000000  
25%      6900.000000  
50%     12100.000000  
75%     18200.000000  
max    230000.000000

#**5.5 Statistical Summary of Object Columns**
A statistical summary includes the count, number of unique values, most frequent value (top), and frequency of the top value for each categorical column.

In [8]:
df.select_dtypes(include='object')

make                model         trim       body transmission  \
0          Kia              Sorento           LX        SUV    automatic   
1          Kia              Sorento           LX        SUV    automatic   
2          BMW             3 Series   328i SULEV      Sedan    automatic   
3        Volvo                  S60           T5      Sedan    automatic   
4          BMW  6 Series Gran Coupe         650i      Sedan    automatic   
...        ...                  ...          ...        ...          ...   
558832     Kia                 K900       Luxury      Sedan          NaN   
558833     Ram                 2500  Power Wagon   Crew Cab    automatic   
558834     BMW                   X5    xDrive35d        SUV    automatic   
558835  Nissan               Altima        2.5 S      sedan    automatic   
558836    Ford                F-150          XLT  SuperCrew    automatic   

                      vin state   color interior  \
0       5xyktca69fg566472    ca   white    black   
1       5xyktca69fg561319    ca   white    beige   
2       wba3c1c51ek116351    ca    gray    black   
3       yv1612tb4f1310987    ca   white    black   
4       wba6b2c57ed129731    ca    gray    black   
...                   ...   ...     ...      ...   
558832  knalw4d4xf6019304    in  silver    black   
558833  3c6td5et6cg112407    wa   white    black   
558834  5uxzw0c58cl668465    ca   black    black   
558835  1n4al3ap0fc216050    ga   white    black   
558836  1ftfw1et2eke87277    ca    gray     gray   

                                                   seller  \
0                                 kia motors america  inc   
1                                 kia motors america  inc   
2                  financial services remarketing (lease)   
3                                 volvo na rep/world omni   
4                  financial services remarketing (lease)   
...                                                   ...   
558832                                   avis corporation   
558833                                    i -5 uhlmann rv   
558834             financial services remarketing (lease)   
558835  enterprise vehicle exchange / tra / rental / t...   
558836                   ford motor credit company llc pd   

                                       saledate  
0       Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1       Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2       Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
3       Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
4       Thu Dec 18 2014 12:30:00 GMT-0800 (PST)  
...                                         ...  
558832  Thu Jul 09 2015 07:00:00 GMT-0700 (PDT)  
558833  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558834  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558835  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)  
558836  Thu May 28 2015 05:30:00 GMT-0700 (PDT)  

[558837 rows x 11 columns]

#**6. Data Cleaning/Preprocessing**
In this section, we assess the integrity and quality of the car sales dataset by examining the completeness, consistency, and data types of each column. We identify missing values, check for correct data types, and verify that the data aligns with expected formats and ranges. This validation ensures that subsequent analyses are based on reliable and accurate data.

## **6.1 Fix Data types**

In this section, we assess we will change the data type of saledate from object to datetime format.

In [15]:
#  = pd.Series()
df['saledate']= pd.to_datetime(df['saledate'], errors = 'coerce', utc= True)
df['saledate'] = pd.to_datetime(df['saledate'].dt.date)
df['saledate']

<ipython-input-15-b48bbbfcdabd>:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['saledate']= pd.to_datetime(df['saledate'], errors = 'coerce', utc= True)


0        2014-12-16
1        2014-12-16
2        2015-01-14
3        2015-01-28
4        2014-12-18
            ...    
558832   2015-07-09
558833   2015-07-08
558834   2015-07-08
558835   2015-07-08
558836   2015-05-27
Name: saledate, Length: 558837, dtype: datetime64[ns]

## **6.2 Handling Missing Values**

We start by identifying and handling any missing values. This ensures the analysis isn't skewed by incomplete data.

In [16]:
# Check for missing values
df.isnull().sum()

year                0
make            10301
model           10399
trim            10651
body            13195
transmission    65352
vin                 4
state               0
condition       11820
odometer           94
color             749
interior          749
seller              0
mmr                38
sellingprice       12
saledate           38
dtype: int64

The dataset has minimal missing values overall, with transmission, trim, and body showing the highest null counts. Key variables like sellingprice, odometer, and saledate are nearly complete and suitable for analysis after basic cleaning.

##**6.3. Dropping All Null rows**

In [17]:
df.dropna()

year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558831  2011     BMW             5 Series         528i      Sedan   
558833  2012     Ram                 2500  Power Wagon   Crew Cab   
558834  2012     BMW                   X5    xDrive35d        SUV   
558835  2015  Nissan               Altima        2.5 S      sedan   
558836  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer  color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0  white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0  white   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0   gray   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0  white   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0   gray   
...             ...                ...   ...        ...       ...    ...   
558831    automatic  wbafr1c53bc744672    fl       39.0   66403.0  white   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0  white   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0  black   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0  white   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0   gray   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558831    brown           lauderdale imports ltd bmw pembrok pines  20300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice   saledate  
0            21500.0 2014-12-16  
1            21500.0 2014-12-16  
2            30000.0 2015-01-14  
3            27750.0 2015-01-28  
4            67000.0 2014-12-18  
...              ...        ...  
558831       22800.0 2015-07-06  
558833       30800.0 2015-07-08  
558834       34000.0 2015-07-08  
558835       11100.0 2015-07-08  
558836       26700.0 2015-05-27  

[472325 rows x 16 columns]

## **6.4. Handle Duplicate Records**

Duplicate records can bias the analysis. We check for and remove any duplicates if found.

In [21]:
# Check for duplicate records
df.duplicated().sum()

np.int64(0)

**Interpretation:** There are no duplicate records in the dataset. This means that we don't need to remove any records before analyzing the dataset.

## **6.5. Standardize Column Names**

To make column names easier to work with, we convert them to lowercase and replace spaces with underscores.


In [11]:
# Check column names to see if they are consistent
df.columns

Index(['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state',
       'condition', 'odometer', 'color', 'interior', 'seller', 'mmr',
       'sellingprice', 'saledate'],
      dtype='object')

##**6.6. Identify Numerical and Categorical Columns**
We categorize columns into numerical and categorical groups to streamline further analysis. This enables targeted handling during cleaning, visualization, and modeling.

In [22]:
# Identify categorical and numerical columns for targeted processing
categorical_cols = df.select_dtypes(include='object').columns.tolist()
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()

In [23]:
# Display categorical columns
list(categorical_cols)

['make',
 'model',
 'trim',
 'body',
 'transmission',
 'vin',
 'state',
 'color',
 'interior',
 'seller']

In [25]:
# Display numerical columns
list(numerical_cols)

['year', 'condition', 'odometer', 'mmr', 'sellingprice']

In [27]:
df['saledate'].dtype

dtype('<M8[ns]')

##**6.7. Standardize Categorical Features and Explore Unique Values**

###**6.7. A. Explore Unique Values in Categorical Features**

In [28]:
# Display unique values for each categorical feature

for col in categorical_cols:
  print(f"Unique values in {col}:")
  print(df[col].unique())
  print("\n")

Unique values in make:
['Kia' 'BMW' 'Volvo' 'Nissan' 'Chevrolet' 'Audi' 'Ford' 'Hyundai' 'Buick'
 'Cadillac' 'Acura' 'Lexus' 'Infiniti' 'Jeep' 'Mercedes-Benz' 'Mitsubishi'
 'Mazda' 'MINI' 'Land Rover' 'Lincoln' 'lincoln' 'Jaguar' 'Volkswagen'
 'Toyota' 'Subaru' 'Scion' 'Porsche' nan 'bmw' 'Dodge' 'FIAT' 'Chrysler'
 'ford' 'Ferrari' 'Honda' 'GMC' 'mitsubishi' 'Ram' 'smart' 'chevrolet'
 'Bentley' 'chrysler' 'pontiac' 'Pontiac' 'Saturn' 'Maserati' 'Mercury'
 'HUMMER' 'landrover' 'cadillac' 'land rover' 'mercedes' 'mazda' 'toyota'
 'lexus' 'gmc truck' 'honda' 'nissan' 'porsche' 'Saab' 'Suzuki' 'dodge'
 'subaru' 'Oldsmobile' 'oldsmobile' 'hyundai' 'jeep' 'Isuzu' 'dodge tk'
 'Geo' 'acura' 'volkswagen' 'suzuki' 'kia' 'audi' 'Rolls-Royce' 'gmc'
 'maserati' 'mazda tk' 'mercury' 'buick' 'hyundai tk' 'mercedes-b' 'vw'
 'Daewoo' 'chev truck' 'ford tk' 'plymouth' 'Plymouth' 'ford truck'
 'Tesla' 'airstream' 'dot' 'Aston Martin' 'Fisker' 'Lamborghini' 'Lotus']


Unique values in model:
['Sorento' '3